rustls/src/msgs/deframer/buffers.rs

```diff
@@ -0,0 +1,115 @@
+/// a `Range` offset within.
+#[derive(Debug)]
+pub(crate) struct Locator {
+    bounds: Range<*const u8>,
+}
+
+impl Locator {
+    pub(crate) fn new(slice: &[u8]) -> Self {
+        Self {
+            bounds: slice.as_ptr_range(),
+        }
+    }
+
+    pub(crate) fn locate(&self, slice: &[u8]) -> Range<usize> {
+        let bounds = slice.as_ptr_range();
+        debug_assert!(self.fully_contains(slice));
+        let start = bounds.start as usize - self.bounds.start as usize;
+        let len = bounds.end as usize - bounds.start as usize;
+        Range {
+            start,
+            end: start + len,
+        }
+    }
+
+    pub(crate) fn fully_contains(&self, slice: &[u8]) -> bool {
+        let bounds = slice.as_ptr_range();
+        bounds.start >= self.bounds.start && bounds.end <= self.bounds.end
+    }
+}
+
+/// Conversion from a `Range` offset to the original slice.
+pub(crate) struct Delocator<'b> {
+    slice: &'b [u8],
+}
+
+impl<'b> Delocator<'b> {
+    pub(crate) fn new(slice: &'b [u8]) -> Self {
+        Self { slice }
+    }
+
+    pub(crate) fn slice_from_range(&'_ self, range: &Range<usize>) -> &'b [u8] {
+        self.slice.get(range.clone()).unwrap()
+    }
+
+    pub(crate) fn locator(self) -> Locator {
+        Locator::new(self.slice)
+    }
+}
+
+/// Reordering the underlying buffer based on ranges.
+pub(crate) struct Coalescer<'b> {
+    slice: &'b mut [u8],
+}
+
+impl<'b> Coalescer<'b> {
+    pub(crate) fn new(slice: &'b mut [u8]) -> Self {
+        Self { slice }
+    }
+
+    pub(crate) fn copy_within(&mut self, from: Range<usize>, to: Range<usize>) {
+        debug_assert!(from.len() == to.len());
+        self.slice.copy_within(from, to.start);
+    }
+
+    pub(crate) fn delocator(self) -> Delocator<'b> {
+        Delocator::new(self.slice)
+    }
+}
+
+/// Accounting structure tracking progress in parsing a single buffer.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct BufferRange {
+    /// Prefix of the buffer that has been processed so far.
+    ///
+    /// `processed` may exceed `discard`, that means we have parsed
+    /// some buffer, but are still using it.  This happens due to
+    /// in-place decryption of incoming records, and in-place
+    /// reassembly of handshake messages.
+    ///
+    /// 0 <= processed <= len
+    processed: usize,
+
+    /// Prefix of the buffer that can be removed.
+    ///
+    /// If `discard` exceeds `processed`, that means we are ignoring
+    /// data without processing it.
+    ///
+    /// 0 <= discard <= len
+    discard: usize,
+}
+
+impl BufferRange {
+    pub(crate) fn add_discard(&mut self, discard: usize) {
+        self.discard += discard;
+    }
+
+    pub(crate) fn add_processed(&mut self, processed: usize) {
+        self.processed += processed;
+    }
+
+    pub(crate) fn take_discard(&mut self) -> usize {
+        self.processed = self
+            .processed
+            .saturating_sub(self.discard);
```

Comment:
it seems slightly non-obvious why `processed` is updated when you take `self.discard`. Maybe worth a clarifying comment?